File: packages/dsa-ts/src/algorithms/searching/binarySearch.ts
/**
 * Binary Search — O(log n). El array DEBE estar ordenado.
 * @returns índice del elemento o -1 si no existe.
 */
export function binarySearch<T>(
  arr: T[],
  target: T,
  comparator: (a: T, b: T) => number = (a, b) => (a < b ? -1 : a > b ? 1 : 0),
): number {
  let left = 0, right = arr.length - 1;

  while (left <= right) {
    const mid = (left + right) >>> 1; // evita overflow
    const cmp = comparator(arr[mid] as T, target);
    if (cmp === 0) return mid;
    if (cmp < 0)   left  = mid + 1;
    else            right = mid - 1;
  }
  return -1;
}

/**
 * Linear Search — O(n). No requiere orden.
 * @returns índice del elemento o -1 si no existe.
 */
export function linearSearch<T>(arr: T[], target: T): number {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === target) return i;
  }
  return -1;
}

File: packages/dsa-py/src/algorithms/searching/binary_search.py
"""
Binary Search — O(log n). Array DEBE estar ordenado.
"""
from typing import TypeVar

T = TypeVar("T")


def binary_search(arr: list[T], target: T) -> int:
    """Devuelve el índice o -1."""
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) >> 1
        if arr[mid] == target:    return mid
        if arr[mid] < target:     left  = mid + 1  # type: ignore[operator]
        else:                     right = mid - 1
    return -1


def linear_search(arr: list[T], target: T) -> int:
    """O(n). No requiere orden."""
    for i, item in enumerate(arr):
        if item == target:
            return i
    return -1
